compiler/src/lnntoamm/Fn.ts

```diff
@@ -0,0 +1,235 @@
+import { LPNode, NulLP } from '../lp';
+import Output from './Amm';
+import opcodes from './opcodes';
+import Scope from './Scope';
+import Stmt, { Assign, Dec, Emit, Exit, FnArg, MetaData, Ref } from './Statement';
+import Type, { FunctionType } from './Types';
+import { TODO } from './util';
+
+// the value is null if the type is to be inferred
+export type Args = {[name: string]: FnArg};
+
+export default class Fn {
+  // null if it's an anonymous fn
+  name: string | null
+  ast: LPNode
+  // the scope this function is defined in is the `par`
+  scope: Scope
+  args: Args
+  retTy: Type
+  body: Stmt[]
+  // not used by this class, but used by Statements
+  metadata: MetaData
+  fnType: FunctionType
+
+  get argNames(): string[] {
+    return Object.keys(this.args);
+  }
+
+  constructor(
+    ast: LPNode,
+    scope: Scope,
+    name: string | null,
+    args: Args,
+    retTy: Type | null,
+    body: Stmt[],
+    metadata: MetaData = null,
+  ) {
+    this.ast = ast;
+    this.scope = scope;
+    this.name = name;
+    this.args = args;
+    for (let argName of Object.keys(this.args)) {
+      if (this.args[argName].ty === null) {
+        this.args[argName].ty = Type.generate();
+      }
+    }
+    this.retTy = retTy !== null ? retTy : Type.generate();
+    this.body = body;
+    this.metadata = metadata !== null ? metadata : new MetaData(scope);
+    this.fnType = new FunctionType(
+      this.name,
+      Object.values(this.args).map(a => a.ty),
+      this.retTy
+    );
+  }
+
+  static fromFunctionsAst(
+    ast: LPNode,
+    scope: Scope,
+    // metadata: MetaData = null,
+  ): Fn {
+    // TODO: inheritance
+    let metadata = new MetaData(scope);
+
+    let work = ast;
```

Comment:
Why this renaming here? I don't think it serves any purpose?

Reply:
I was reassigning `work` in a previous iteration so that I can keep the original `ast`, but it's no longer needed